File: OGLEngine/OBJLoader.m
//
//  OBJLoader.m
//  jj
//
//  Created by Maciej Chmielewski on 16.02.2016.
//  Copyright © 2016 AppUnite. All rights reserved.
//

#import "OBJLoader.h"

@interface OBJLoadingResult ()

@property (nonatomic, assign, readwrite) OBJInfo objInfo;
@property (nonatomic, assign, readwrite) GLfloat **positions;
@property (nonatomic, assign, readwrite) GLfloat **texels;
@property (nonatomic, assign, readwrite) GLfloat **normals;
@property (nonatomic, assign, readwrite) GLint **faces;

@end

@implementation OBJLoadingResult

@end


@implementation OBJLoader

+ (OBJ *)objFromFileNamed:(NSString *)fileName {
    OBJLoadingResult *result = [OBJLoader loadOBJNamed:fileName];
    return [OBJLoader objFromStrideCollection:[OBJLoader strideDataFromOBJLoadingResult:result]
                             objLoadingResult:result];
}

+ (OBJ *)objFromStrideCollection:(StrideCollection *)strideCollection objLoadingResult:(OBJLoadingResult *)objLoadingResult {
    // Indices
    GLuint *indices = malloc(strideCollection.indices.count * sizeof(GLuint));
    for (int i=0; i<strideCollection.indices.count; i++) {
        indices[i] = [strideCollection.indices[i] unsignedIntValue];
    }
    GLIntArray *indicesArray = [GLIntArray new];
    indicesArray.data = indices;
    indicesArray.count = strideCollection.indices.count;
    
    // Vertex data
    GLfloat *positions = malloc(strideCollection.array.count * 3 * sizeof(GLfloat));
    GLfloat *texels = malloc(strideCollection.array.count * 2 * sizeof(GLfloat));
    GLfloat *normals = malloc(strideCollection.array.count * 3 * sizeof(GLfloat));
    OBJInfo structIterator = {0};
    for (StrideData *stride in strideCollection.array) {
        positions[structIterator.positions ++] = stride.positions[0];
        positions[structIterator.positions ++] = stride.positions[1];
        positions[structIterator.positions ++] = stride.positions[2];
        
        texels[structIterator.texels ++] = stride.texels[0];
        texels[structIterator.texels ++] = stride.texels[1];
        
        normals[structIterator.normals ++] = stride.normals[0];
        normals[structIterator.normals ++] = stride.normals[1];
        normals[structIterator.normals ++] = stride.normals[2];
    }
    
    GLFloatArray *positionsArray = [GLFloatArray new];
    positionsArray.data = positions;
    positionsArray.count = structIterator.positions;
    
    GLFloatArray *texelsArray = [GLFloatArray new];
    texelsArray.data = texels;
    texelsArray.count = structIterator.texels;
    
    GLFloatArray *normalsArray = [GLFloatArray new];
    normalsArray.data = normals;
    positionsArray.count = structIterator.positions;
    
    return [[OBJ alloc] initWithIndices:indicesArray
                              positions:positionsArray
                                 texels:texelsArray
                                normals:normalsArray];
}

+ (StrideCollection *)strideDataFromOBJLoadingResult:(OBJLoadingResult *)objLoadingResult {
    StrideCollection *strides = [StrideCollection new];
    
    for (int i=0; i<objLoadingResult.objInfo.faces; i++) {
        [strides addObject:[[StrideData alloc] initWithPositionIndex:objLoadingResult.faces[i][0]
                                                          texelIndex:objLoadingResult.faces[i][1]
                                                         normalIndex:objLoadingResult.faces[i][2]
                                                           positions:objLoadingResult.positions
                                                              texels:objLoadingResult.texels
                                                             normals:objLoadingResult.normals]];
        [strides addObject:[[StrideData alloc] initWithPositionIndex:objLoadingResult.faces[i][3]
                                                          texelIndex:objLoadingResult.faces[i][4]
                                                         normalIndex:objLoadingResult.faces[i][5]
                                                           positions:objLoadingResult.positions
                                                              texels:objLoadingResult.texels
                                                             normals:objLoadingResult.normals]];
        [strides addObject:[[StrideData alloc] initWithPositionIndex:objLoadingResult.faces[i][6]
                                                          texelIndex:objLoadingResult.faces[i][7]
                                                         normalIndex:objLoadingResult.faces[i][8]
                                                           positions:objLoadingResult.positions
                                                              texels:objLoadingResult.texels
                                                             normals:objLoadingResult.normals]];
    }
    return strides;
}

+ (OBJLoadingResult *)loadOBJNamed:(NSString *)name {
    OBJLoadingResult *result = [OBJLoadingResult new];
    
    // read file
    NSString *filePath = [[NSBundle mainBundle] pathForResource:name ofType:@"obj"];
    NSString *payload = [NSString stringWithContentsOfFile:filePath];
    NSArray<NSString *> *lines = [payload componentsSeparatedByString:@"\n"];
    
    // read info
    OBJInfo objInfo = {0};
    for (NSString *line in lines) {
        if ([line hasPrefix:@"v "]) {
            objInfo.positions ++;
        } else if ([line hasPrefix:@"vn"]) {
            objInfo.normals ++;
        } else if ([line hasPrefix:@"vt"]) {
            objInfo.texels ++;
        } else if ([line hasPrefix:@"f "]) {
            objInfo.faces ++;
        }
    }
    objInfo.vertices = objInfo.faces * 3;
    result.objInfo = objInfo;
    
    // init arrays
    result.positions = malloc(objInfo.positions * sizeof(GLfloat *));
    for (int i=0; i<=objInfo.positions - 1; i++) {
        result.positions[i] = malloc(3 * sizeof(GLfloat));
    }
    result.texels = malloc(objInfo.texels * sizeof(GLfloat *));
    for (int i=0; i<=objInfo.texels - 1; i++) {
        result.texels[i] = malloc(2 * sizeof(GLfloat));
    }
    result.normals = malloc(objInfo.normals * sizeof(GLfloat *));
    for (int i=0; i<=objInfo.normals - 1; i++) {
        result.normals[i] = malloc(3 * sizeof(GLfloat));
    }
    result.faces = malloc(objInfo.faces * sizeof(GLint *));
    for (int i=0; i<=objInfo.faces - 1; i++) {
        result.faces[i] = malloc(9 * sizeof(GLint));
    }
    
    
    OBJInfo structIterator = {0};
    for (NSString *line in lines) {
        if ([line hasPrefix:@"v "]) {
            result.positions[structIterator.positions] = [OBJLoader floatsFromString:line];
            structIterator.positions ++;
        } else if ([line hasPrefix:@"vn"]) {
            result.normals[structIterator.normals] = [OBJLoader floatsFromString:line];
            structIterator.normals ++;
        } else if ([line hasPrefix:@"vt"]) {
            result.texels[structIterator.texels] = [OBJLoader floatsFromString:line];
            structIterator.texels ++;
        } else if ([line hasPrefix:@"f "]) {
            result.faces[structIterator.faces] = [OBJLoader facesFromString:line];
            structIterator.faces ++;
        }
    }
    return result;
}

+ (GLint *)facesFromString:(NSString *)string {
    string = [string stringByReplacingOccurrencesOfString:@"f" withString:@""];
    string = [string stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSArray *triangles = [string componentsSeparatedByString:@" "];
    
    GLint *result = malloc(9 * sizeof(GLint));
    unsigned int index = 0;
    for (NSString *triangle in triangles) {
        NSArray<NSString *> *components = [triangle componentsSeparatedByString:@"/"];
        for (NSString *component in components) {
            result[index] = [component intValue];
            index ++;
        }
    }
    
    return result;
}

+ (GLfloat *)floatsFromString:(NSString *)string {
    NSArray <NSString *> *floatStrings = [string componentsSeparatedByString:@" "];
    NSMutableArray *mFloatStrings = [floatStrings mutableCopy];
    [mFloatStrings removeObjectAtIndex:0];
    floatStrings = mFloatStrings;
    GLfloat *floats = malloc(floatStrings.count * sizeof(GLfloat));
    for (int i=0; i<=floatStrings.count - 1; i++) {
        floats[i] = [floatStrings[i] floatValue];
    }
    return floats;
}

@end


@implementation StrideData

- (instancetype)initWithPositionIndex:(NSUInteger)positionIndex
                           texelIndex:(NSUInteger)texelIndex
                          normalIndex:(NSUInteger)normalIndex
                            positions:(GLfloat **)positions
                               texels:(GLfloat **)texels
                              normals:(GLfloat **)normals {
    self = [super init];
    if (self) {
        self.positionIndex = positionIndex;
        self.texelIndex = texelIndex;
        self.normalIndex = normalIndex;
        self.positions = positions[positionIndex - 1];
        self.texels = texels[texelIndex - 1];
        self.normals = normals[normalIndex - 1];
    }
    return self;
}

- (BOOL)isEqual:(id)object {
    if ([super isEqual:object]) {
        return YES;
    }
    if ([object isKindOfClass:[StrideData class]]) {
        return
        ((self.positionIndex == ((StrideData *)object).positionIndex) &&
         (self.texelIndex == ((StrideData *)object).texelIndex) &&
         (self.normalIndex == ((StrideData *)object).normalIndex));
    }
    return NO;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"%d-%d-%d", self.positionIndex, self.texelIndex, self.normalIndex];
}

@end


@implementation StrideCollection

- (NSMutableArray<StrideData *> *)array {
    if (!_array) {
        _array = [NSMutableArray new];
    }
    return _array;
}

- (NSMutableArray<NSNumber *> *)indices {
    if (!_indices) {
        _indices = [NSMutableArray new];
    }
    return _indices;
}

- (void)addObject:(StrideData *)stride {
    if ([self.array containsObject:stride]) {
        [self.indices addObject:@([self.array indexOfObject:stride] + 1)];
        return;
    }
    [self.array addObject:stride];
    [self.indices addObject:@(self.array.count)];
}

@end

File: OGLEngine/OBJLoader.swift
//
//  OBJLoader.swift
//  OGLEngine
//
//  Created by Maciej Chmielewski on 24.03.2016.
//  Copyright © 2016 MaciejCh. All rights reserved.
//

import Foundation
import GLKit

struct Vec3 {
    let x: Float
    let y: Float
    let z: Float
    
    init(line: String) {
        let c = line.componentsSeparatedByString(" ")
        self.x = Float(c[1])!
        self.y = Float(c[2])!
        self.z = Float(c[3])!
    }
    
    init(x: Float, y: Float, z: Float) {
        self.x = x
        self.y = y
        self.z = z
    }
    
    func sub(vec3: Vec3) -> Vec3 {
        return Vec3(x: self.x - vec3.x,
                    y: self.y - vec3.y,
                    z: self.z - vec3.z)
    }
    
    func scale(s: Float) -> Vec3 {
        return Vec3(x: self.x * s,
                    y: self.y * s,
                    z: self.z * s)
    }
}

struct Vec2 {
    let u: Float
    let v: Float
    
    init(line: String) {
        let c = line.componentsSeparatedByString(" ")
        self.u = Float(c[1])!
        self.v = Float(c[2])!
    }
    
    init(u: Float, v: Float) {
        self.u = u
        self.v = v
    }
    
    func sub(vec2: Vec2) -> Vec2 {
        return Vec2(u: self.u - vec2.u, v: self.v - vec2.v)
    }
}

struct Vertex {
    let indexIdentifier: VertexIndices
    let position: Vec3
    let texel: Vec2
    let normal: Vec3
    let tangent: Vec3?
}

func normalVersor(v1: Vec3, v2: Vec3) -> Vec3 {
    let gv1 = GLKVector3Make(v1.x, v1.y, v1.z)
    let gv2 = GLKVector3Make(v2.x, v2.y, v2.z)
    let res = GLKVector3Normalize(GLKVector3CrossProduct(gv1, gv2))
    return Vec3(x: res.x, y: res.y, z: res.z)
}

func dot(v1: Vec3, gv2: GLKVector3) -> Float {
    let gv1 = GLKVector3Make(v1.x, v1.y, v1.z)
    return GLKVector3DotProduct(gv1, gv2)
}


public func ==(lhs: VertexIndices, rhs: VertexIndices) -> Bool {
    if (lhs.p != rhs.p) {
        return false
    }
    if (lhs.t != rhs.t) {
        return false
    }
    if (lhs.n != rhs.n) {
        return false
    }
    return true
}

public struct VertexIndices: Equatable {
    let p: Int
    let t: Int
    let n: Int
    
    init(line: String) {
        let c = line.componentsSeparatedByString("/")
        self.p = Int(c[0])!
        self.t = Int(c[1])!
        self.n = Int(c[2])!
    }
}

struct FaceIndices {
    let v1: VertexIndices
    let v2: VertexIndices
    let v3: VertexIndices
    
    init(line: String) {
        let c = line.componentsSeparatedByString(" ")
        self.v1 = VertexIndices(line: c[1])
        self.v2 = VertexIndices(line: c[2])
        self.v3 = VertexIndices(line: c[3])
    }
}

struct Face {
    let v1: Vertex
    let v2: Vertex
    let v3: Vertex
}

class OBJLoader : NSObject {
    
    class func objFromFileNamed(fileName: String) -> OBJ {
        debugPrint(" will load obj")
        let filePath: String = NSBundle.mainBundle().pathForResource("3dAssets/meshes/" + fileName, ofType: "obj")!
        
        var payload = ""
        do {
            try payload = String(contentsOfFile: filePath)
        } catch _ {
            assert(false, "Could not read file")
        }
        let lines: [String] = payload.componentsSeparatedByString("\n")
        
        var positions: [Vec3] = []
        var texels: [Vec2] = []
        var normals: [Vec3] = []
        var facesIndices: [FaceIndices] = []
        
        for line in lines {
            if line.startIndex.distanceTo(line.endIndex) < 2 {
                continue
            }
            switch line.substringToIndex(line.startIndex.advancedBy(2)) {
            case "v ": positions.append(Vec3(line: line))
            case "vt": texels.append(Vec2(line: line))
            case "vn": normals.append(Vec3(line: line))
            case "f ": facesIndices.append(FaceIndices(line: line))
            default: break
            }
        }
        debugPrint("  did read file")
        let indicesToVertex = { (indices: VertexIndices) -> Vertex in
            return Vertex(indexIdentifier: indices,
                          position: positions[indices.p - 1],
                          texel: texels[indices.t - 1],
                          normal: normals[indices.n - 1],
                          tangent: nil)
        }
        
        let faces = facesIndices.map{
            Face(v1: indicesToVertex($0.v1), v2: indicesToVertex($0.v2), v3: indicesToVertex($0.v3))
        }
        
        
        let tangentFaces = faces.map { (let face) -> Face in
            let independent = Vec3(x: 0, y: 0, z: 1)
            let tangent1 = normalVersor(face.v1.normal, v2: independent)
            let tangent2 = normalVersor(face.v2.normal, v2: independent)
            let tangent3 = normalVersor(face.v3.normal, v2: independent)
                return Face(
                    v1: Vertex(
                        indexIdentifier: face.v1.indexIdentifier,
                        position: face.v1.position,
                        texel: face.v1.texel,
                        normal: face.v1.normal,
                        tangent: tangent1),
                    v2: Vertex(
                        indexIdentifier: face.v2.indexIdentifier,
                        position: face.v2.position,
                        texel: face.v2.texel,
                        normal: face.v2.normal,
                        tangent: tangent2),
                    v3: Vertex(
                        indexIdentifier: face.v3.indexIdentifier,
                        position: face.v3.position,
                        texel: face.v3.texel,
                        normal: face.v3.normal,
                        tangent: tangent3))
        }
        debugPrint("  did calculate tangents")
        
        let verticesInOrder = tangentFaces.map{[$0.v1, $0.v2, $0.v3]}.stomp()
        debugPrint("  did sort vertices")
        
        var vertexDrawOrder: [Int] = []
        var drawnVertices: [VertexIndices] = []
        
//        let count = verticesInOrder.count
//        var i = 0
        let nonRepeatingVerticesInOrder = verticesInOrder.filter{
//            debugPrint("\(i)/\(count)")
//            i += 1;
//            if let index = drawnVertices.indexOf($0.indexIdentifier) {
//                vertexDrawOrder.append(index)
//                return false
//            } else {
                drawnVertices.append($0.indexIdentifier)
                vertexDrawOrder.append(drawnVertices.count - 1)
                return true
//            }
        }
        debugPrint("  did calculate indices")
        let positionsArray = nonRepeatingVerticesInOrder.map{$0.position}.map{[$0.x, $0.y, $0.z]}.stomp()
        let texelsArray = nonRepeatingVerticesInOrder.map{$0.texel}.map{[$0.u, $0.v]}.stomp()
        let normalsArray = nonRepeatingVerticesInOrder.map{$0.normal}.map{[$0.x, $0.y, $0.z]}.stomp()
        let tangentsArray = nonRepeatingVerticesInOrder.map{$0.tangent}.map{[$0!.x, $0!.y, $0!.z]}.stomp()
        
        debugPrint("did load obj")
        return OBJ(indices: vertexDrawOrder, positions: positionsArray, texels: texelsArray, normals: normalsArray, tangents: tangentsArray)
    }
    
}

func calculateTangents(v0: Vertex, v1: Vertex, v2: Vertex) -> Vec3 {
    var delatPos1 = v1.position.sub(v0.position)
//    Vector3f delatPos1 = Vector3f.sub(v1.getPosition(), v0.getPosition(), null);
    var delatPos2 = v2.position.sub(v0.position)
//    Vector3f delatPos2 = Vector3f.sub(v2.getPosition(), v0.getPosition(), null);
    
//    Vector2f uv0 = textures.get(v0.getTextureIndex());
//    Vector2f uv1 = textures.get(v1.getTextureIndex());
//    Vector2f uv2 = textures.get(v2.getTextureIndex());
    let deltaUv1 = v1.texel.sub(v0.texel)
//    Vector2f deltaUv1 = Vector2f.sub(uv1, uv0, null);
    let deltaUv2 = v2.texel.sub(v0.texel)
//    Vector2f deltaUv2 = Vector2f.sub(uv2, uv0, null);
//    
//    float r = 1.0f / (deltaUv1.x * deltaUv2.y - deltaUv1.y * deltaUv2.x);
    let r = 1 / (deltaUv1.u * deltaUv2.v - deltaUv1.v * deltaUv2.u)
//    delatPos1.scale(deltaUv2.y);
    delatPos1 = delatPos1.scale(deltaUv2.v)
//    delatPos2.scale(deltaUv1.y);
    delatPos2 = delatPos2.scale(deltaUv1.v)
//    Vector3f tangent = Vector3f.sub(delatPos1, delatPos2, null);
    var tangent = delatPos1.sub(delatPos2)
//    tangent.scale(r);
    tangent = tangent.scale(r)
//    v0.addTangent(tangent);
//    v1.addTangent(tangent);
//    v2.addTangent(tangent);
    
    return tangent
}

func rotationFromVector(formVector: GLKVector3, toVector: GLKVector3) -> GLKMatrix3 {
    let dot = GLKVector3DotProduct(formVector, toVector);
    switch dot {
    case 1: return GLKMatrix3Identity
    case -1: return GLKMatrix3MakeRotation(Float(M_PI), 1, 0, 0)
    default: break
    }
    
    let axis = GLKVector3Normalize(GLKVector3CrossProduct(formVector, toVector));
    let radians = acosf(dot);
    let mat = GLKMatrix3RotateWithVector3(GLKMatrix3Identity, radians, axis);
    return mat;
}

extension Array where Element : CollectionType {
    func stomp() -> [Element.Generator.Element] {
        var result: [Element.Generator.Element] = []
        for element in self {
            result.appendContentsOf(element)
        }
        return result
    }
}
